src/config.rs

```diff
@@ -0,0 +1,104 @@
+use std::{collections::HashMap, ffi::OsStr, fs};
+
+use common::extensions::*;
+use microtemplate::render;
+use serde::{Deserialize, Deserializer};
+use uniquote::Quote;
+
+use crate::paths::PATHS;
+
+#[derive(Clone, Default, Debug)]
+pub struct ConfigContext;
+
+impl microtemplate::Context for ConfigContext {
+    fn get_field(&self, field_name: &str) -> &str {
+        match field_name {
+            "game_dir" => PATHS.game_str.as_str(),
+            _ => "",
+        }
+    }
+}
+
+#[derive(Debug, Deserialize)]
+pub struct ModConfig {
+    #[serde(default)]
+    pub(crate) args: HashMap<String, String>,
+    #[serde(default)]
+    pub(crate) tasks: Vec<Task>,
+}
+
+fn default_as_false() -> bool {
+    false
+}
+
+fn default_as_true() -> bool {
+    true
+}
+
+#[derive(Debug, Deserialize, Clone)]
+pub struct Task {
+    pub(crate) command: String,
+    #[serde(default = "default_as_false")]
+    pub(crate) terminate_on_errors: bool,
+    #[serde(default = "default_as_true")]
+    pub(crate) no_window: bool,
+    #[serde(default)]
+    pub(crate) args: Vec<String>,
+    #[serde(flatten, deserialize_with = "deserialize_render_map")]
+    pub(crate) values: HashMap<String, String>,
+}
+
+fn deserialize_render_map<'de, D>(deserializer: D) -> Result<HashMap<String, String>, D::Error>
+where
+    D: Deserializer<'de>,
+{
+    let hash_map = HashMap::<String, String>::deserialize(deserializer)?;
+    Ok(hash_map
+        .into_iter()
+        .map(|(key, value)| (key, render(value.as_str(), ConfigContext {})))
+        .collect())
+}
+
+impl microtemplate::Context for Task {
+    fn get_field(&self, field_name: &str) -> &str {
+        let result = if field_name == "game_dir" {
+            PATHS.game_str.as_str()
+        } else if let Some(val) = self.values.get(field_name) {
+            val.as_str()
+        } else {
+            ""
+        };
+        log::debug!("Task {} key {} = {}", self.command, field_name, result);
+
+        result
+    }
+}
```

Comment:
same as above, I think the code should generally not implement `microtemplate::Context`, it's not the appropriate place to put app logic in

Reply:
Okay, I can just copy (or reference?) game_dir instead of looking it up like that. Slowly getting my mind around borrows and references. It's an interesting set of abstractions Rust uses. It blurs a lot of the lines I have in my mind between low level assembly/C/C++ pointers and higher-level dynamic garbage-collected language's references and objects.

I like it. But old habits from different models keep tripping me up!

Reply:
The simplest solution is to just shove a `PathBuf` into the struct. The way `microtemplate` works might then force you to do some redundant `clone`s, which would be a bit inefficient, but I think that'd be fine. Alternatively you could try to borrow it as a `Path`, but you don't have to if it proves difficult.

Reply:
Oh, that's the `Task` `microtemplate::Context`. I have no idea how to implement that one using the derive macro. I don't think it can be. The keys are dynamic, they are defined in each task's toml file. So they can't be known to name as struct fields at compile time.

I can pre-set the game_dir value and remove that lookup. But the HashMap's `.get` call is not something microtemplate's derive macro supports.

Reply:
What's the purpose of the hash map? I don't think it's being used anywhere, is the idea that the user can provide custom template parameters? If that's what you want then microtemplate would indeed not work well with this approach. Is that actually needed though? I'd rather just get rid of the hash map and only prepopulate it with known parameters like game_dir.

Reply:
> I think that's cleaner than having special cases for keys inside the implementations of Context.

Yup, that's pretty much what I meant with:

> I can pre-set the game_dir value and remove that lookup.

I'll look at taking it a step further with the custom DynamicKeys type. I'll see if I find a clean way to implement it around the deserialization and other processing.

> is the idea that the user can provide custom template parameters?

yes

> Is that actually needed though?

Not sure. When it comes to UX, ergonomics, and discoverability, I have trouble answering definitively "what is needed."

Here is the example toml I provided for redscript/scc. (Modified for clarity in this discussion.)
```toml
[[tasks]]
command = "scc"
args = ["-compile", "{path}", "-customCacheDir", "{custom_cache_dir}"]
path = "{game_dir}/r6/scripts"
custom_cache_dir = "{game_dir}/r6/cache/modded"
```
If you look at `args` you can see how `path` and `custom_cache_dir` are custom keys being defined for this task.

Alternatively it could be done plain like:
```toml
[[tasks]]
command = "scc"
args = ["-compile", "{game_dir}/r6/scripts", "-customCacheDir", "{game_dir}/r6/cache/modded"]
```

But I'm not sure that simplification is worthwhile for the difference in clarity. Other alternatives could be something like:
```toml
[[tasks]]
command = "scc"
"-compile" ="{game_dir}/r6/scripts"
"-customCacheDir" = "{game_dir}/r6/cache/modded"
```
Or, to avoid the quoting of keys:
```toml
[[tasks]]
command = "scc"
argument_prefix = "-"
compile ="{game_dir}/r6/scripts"
customCacheDir = "{game_dir}/r6/cache/modded"
```

Thoughts or preferences?

Reply:
> 1 and 2 are OK.

Okay, I'll continue with 1 and see if I can clean up the implementation to an acceptable level. It's easy enough to remove if 2 ends up being the final decision.

Reply:
Oh, additional thought.

Side notes/commentary:
* I'm going to write this out as if you don't know the fine details of toml syntax. You may know it far better than I do. I think it works better to try to explain more clearly than less.
* I'm focusing on the toml format, since it explains and drives a lot of my decisions. Knowing the intended goal may help you better understand and evaluate my bizarre and chaotic work 😄 

Now back to the thought, the topic of this comment -
We could modify the syntax to something like:
```toml
[tasks.redscript]
command = "scc"
args = ["-compile", "{path}", "-customCacheDir", "{custom_cache_dir}"]

[tasks.redscript.arg_values]
path = "{game_dir}/r6/scripts"
custom_cache_dir = "{game_dir}/r6/cache/modded"
```

Or doing similar with array syntax combined with subtable syntax:
```toml
[[tasks]]
command = "scc"
args = ["-compile", "{path}", "-customCacheDir", "{custom_cache_dir}"]

[tasks.arg_values]
path = "{game_dir}/r6/scripts"
custom_cache_dir = "{game_dir}/r6/cache/modded"
```
Since a subtable in toml applies to the most recently defined array element, those argument values go in the above [[task]] element.

We can also use alternative equivalent syntax like dotted keys:
```toml
[tasks.redscript]
command = "scc"
args = ["-compile", "{path}", "-customCacheDir", "{custom_cache_dir}"]
arg_values.path = "{game_dir}/r6/scripts"
arg_values.custom_cache_dir = "{game_dir}/r6/cache/modded"
```
or an inline table:
```toml
arg_values = { path = "{game_dir}/r6/scripts", custom_cache_dir = "{game_dir}/r6/cache/modded" }
```
Both of which would deserialize identically to the earlier examples.

Reply:
I'd probably name it something more explicit like `template_args` or `substitutions` or something like that.
The syntax I'd prefer is:
```toml
[[tasks]]
command = "scc"
args = ["-compile", "{path}", "-customCacheDir", "{custom_cache_dir}"]
substitutions.path = "{game_dir}/r6/scripts"
substitutions.custom_cache_dir = "{game_dir}/r6/cache/modded"
```
Splitting substitutions to another subtable looks nice but it could be surprising to someone not deeply familiar with TOML (especially the fact that the order matters).